### PR TITLE
Move negative index checking to common.py - Fix issue 97365

### PR DESCRIPTION
**Pull request by guilhermeleobas**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108690

Fixes https://github.com/pytorch/pytorch/issues/97365

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

-- for meta-internal --
@diff-train-skip-merge
this PR is being re-landed as [D50755796](https://www.internalfb.com/intern/diff/D50755796/)

### CODE REVIEW

File: torch/_inductor/codegen/common.py [hunk 1]
```diff
@@ -917,9 +917,34 @@ def inner(*args, **kwargs):
                 return inner
 
             @staticmethod
-            def indirect_indexing(index_var, size, check=True):
+            def indirect_indexing(var, size, check=True):
                 # Skip CSE since this doesn't return an expression
```
Comment (lezcano) [inline torch/_inductor/codegen/common.py:921]: nit. Move this to the bottom.

File: test/inductor/test_torchinductor.py [hunk 2]
```diff
@@ -787,6 +787,66 @@ def repeat(x, n):
         self.assertEqual(expect, actual)
         self.assertEqual(actual, repeat(x, 3))
 
+    @skipIfRocm
+    def test_neg_index(self):
+        def test(fn, inps, has_assert: bool, has_wrapping=True):
+            for dynamic in (True, False):
+                fn_opt = torch.compile(dynamic=dynamic)(fn)
+                if self.device == "cpu":
+                    code = run_and_get_cpp_code(fn_opt, *inps)
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:796]: Add the equivalent asserts in CPU please.

**Review by lezcano — COMMENTED** [review]
Nice! Just two very small things.

File: test/inductor/test_torchinductor.py (hunk 2 above)
Comment (jgong5) [reply test/inductor/test_torchinductor.py:796]: @guilhermeleobas I saw you are adding assertions in a separate PR https://github.com/pytorch/pytorch/pull/108910. Are you going to add the checkers there too?

Comment (guilhermeleobas) [reply test/inductor/test_torchinductor.py:796]: Yes

File: torch/_inductor/codegen/cpp.py [hunk 3]
```diff
@@ -1209,8 +1210,38 @@ def scale_index_with_offset(
...
         return new_index
 
-    @staticmethod
-    def indirect_indexing(index_var, size, check=True):
+    def index_to_str(self, index: sympy.Expr) -> str:
+        """
+        Convert an index expr to a string that can be used in triton code.
+        e.g. a sympy expression "s2" may actually appear as "ks1" in the cpp kernel.
+        """
+        return cexpr(self.rename_indexing(index))
```
Comment (lezcano) [inline torch/_inductor/codegen/cpp.py:1218]: Rather than duplicating the code, move it to `common.py`

File: torch/_inductor/codegen/cpp.py [hunk 4]
```diff
@@ -1209,8 +1210,38 @@ def scale_index_with_offset(
...
+
+    def indirect_indexing(self, index_var, size, check=True):
+        generate_assert = (
+            check or config.debug_index_asserts
+        ) and config.cpp.assert_indirect_indexing
```
Comment (lezcano) [inline torch/_inductor/codegen/cpp.py:1223]: As discussed before, let's just have one way to deactivate the asserts in both CPU and CUDA.

Then, you can lift all this code to `common` rather than duplicating it, and just have an op in each of the codegen's that tells you whether to use `TORCH_CHECK` or `tl.device_assert`.

File: torch/_inductor/codegen/common.py [hunk 5]
```diff
@@ -917,9 +975,72 @@ def inner(*args, **kwargs):
...
+                        mask_vars = set(var.mask_vars)
+                        if self._load_mask:
+                            mask_vars.add(self._load_mask)
+
+                        if mask_vars:
+                            mask = (
+                                f"{list(mask_vars)[0]}"
+                                if len(mask_vars) == 1
+                                else f"({' & '.join(str(v) for v in mask_vars)})"
+                            )
```
Comment (lezcano) [inline torch/_inductor/codegen/common.py:1018]: All the mask logic is triton-only, so it should be in a per-backend aux function. For the same reason `_load_mask`  should just be in the triton kernel

File: test/inductor/test_torchinductor.py [hunk 6]
```diff
@@ -787,6 +787,79 @@ def repeat(x, n):
...
+                    lines = code.split("\n")
+                    found = False
+                    for line in lines:
+                        # Line will look something like this:
+                        # TORCH_CHECK((0 <= tmp5) & (tmp5 < ks1), error_msg)
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:801]: You don't want to match this one, you want to match op.where, that is, the ternary operator when it's not vectorised and `blendv` when it is.

File: torch/_inductor/codegen/common.py (hunk 5 above)
Comment (guilhermeleobas) [reply torch/_inductor/codegen/common.py:1018]: The cpp kernel seems to use `_load_mask` in some places as well. I'll keep it on `common.py` but move the code above to triton.

File: test/inductor/test_torchinductor.py (hunk 6 above)
Comment (guilhermeleobas) [reply test/inductor/test_torchinductor.py]: Is `blendv` a c++ function? I don't see it in any of the C++ generated code

Comment (lezcano) [reply test/inductor/test_torchinductor.py]: It's an [intel intrinsic](https://www.intel.com/content/www/us/en/docs/cpp-compiler/developer-guide-reference/2021-8/mm256-blendv-epi8.html). It's generated if the code is vectorised (it may not be the case in that test, idk).

File: test/inductor/test_torchinductor.py [hunk 7]
```diff
@@ -800,6 +800,76 @@ def repeat(x, n):
...
+                if self.device == "cpu":
+                    code = run_and_get_cpp_code(fn_opt, *inps)
+                    lines = code.split("\n")
+                    found = False
+                    if "Vectorize" not in code:
+                        for line in lines:
+                            # match ternary operator
+                            if "?" in line:
+                                pos = line.find("?")
+                                found |= ":" in line[pos + 1 :]
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:817]: Just regex this

Comment (lezcano) [reply test/inductor/test_torchinductor.py:817]: Also, are there any other ?'s in the code, or could you simply try to find a ? in the code?

File: test/inductor/test_torchinductor.py [hunk 8]
```diff
@@ -800,6 +800,79 @@ def repeat(x, n):
...
+                fn_opt = torch.compile(dynamic=dynamic)(fn)
+                if self.device == "cpu":
+                    code = run_and_get_cpp_code(fn_opt, *inps)
+                    lines = code.split("\n")
+                    found = False
+                    if "Vectorize" not in code:
+                        for line in lines:
+                            # match ternary operator
+                            pattern = r"\?.*:"
+                            if re.findall(pattern, line):
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:816]: I mean, you can do that in the whole code in one go, right?

File: torch/_inductor/codegen/cpp.py [hunk 9]
```diff
@@ -1215,9 +1214,12 @@ def scale_index_with_offset(
         new_index = sympy_subs(index, replacement)
         return new_index
 
-    @staticmethod
-    def indirect_indexing(index_var, size, check=True):
-        return sympy_symbol(str(index_var))
+    def index_to_str(self, index: sympy.Expr) -> str:
+        """
+        Convert an index expr to a string that can be used in triton code.
```
Comment (jgong5) [inline torch/_inductor/codegen/cpp.py:1219]: nit:
```suggestion
        Convert an index expr to a string that can be used in cpp code.
```

File: test/inductor/test_torchinductor.py (hunk 8 above)
Comment (guilhermeleobas) [reply test/inductor/test_torchinductor.py:816]: Yes, right.

File: test/inductor/test_torchinductor.py [hunk 10]
```diff
@@ -800,6 +800,74 @@ def repeat(x, n):
...
 
+    @skipIfRocm
+    def test_neg_index(self):
+        def test(fn, inps, has_assert: bool, has_wrapping=True):
+            for dynamic in (True, False):
+                fn_opt = torch.compile(dynamic=dynamic)(fn)
+                if self.device == "cpu":
+                    code = run_and_get_cpp_code(fn_opt, *inps)
+                    found = False
+                    if "Vectorize" not in code:
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:811]: What happens if we find Vectorize in the code? If none of these vectorize, you can simply remove this if and move on

File: torch/_inductor/codegen/cpp.py (hunk 3 above)
Comment (lezcano) [reply torch/_inductor/codegen/cpp.py]: fair enough.

File: torch/_inductor/codegen/cpp.py [hunk 11]
```diff
@@ -1192,7 +1192,6 @@ def __init__(self, args, num_threads):
         self.poststores = IndentedBuffer()
         self.num_threads = num_threads  # num_threads the kernel specialized for
         self.reduction_omp_dec: Dict[Tuple[str, str], str] = {}
-        self._load_mask = None
```
Comment (peterbell10) [inline torch/_inductor/codegen/cpp.py:1195]: The load mask is relevant to C++ kernels. When you have vectorized loads, it can mask out individual elements in a vector.

File: torch/_inductor/codegen/cpp.py [hunk 12]
```diff
@@ -1410,6 +1412,10 @@ def codegen_loops(self, code, worksharing):
         loop_nest = LoopNestWithSplit.build(self)
         self.codegen_loops_impl(loop_nest, code, worksharing)
 
+    @property
+    def assert_function(self):
+        return "TORCH_CHECK"
```
Comment (peterbell10) [inline torch/_inductor/codegen/cpp.py:1417]: You will probably need a vectorized version as well which checks all elements of the vector are true.

File: torch/_inductor/codegen/cpp.py (hunk 11 above)
Comment (guilhermeleobas) [reply torch/_inductor/codegen/cpp.py:1195]: I've moved this variable to `common.py::Kernel` as both `CppKernel` and `TritonKernel` had a version of it

Comment (peterbell10) [reply torch/_inductor/codegen/cpp.py:1195]: Okay but the point still stands that `load_mask` shouldn't be a no-op on CPU.

File: test/inductor/test_torchinductor.py (hunk 10 above)
Comment (jgong5) [reply test/inductor/test_torchinductor.py:811]: It is possible to have `Vectorized` in the code when there is indirect indexing. The load is conducted with scalar in a loop into the vector and the compute is applied on the vectors. I'm wondering how negative indexing is handled here. @guilhermeleobas is your test generating "Vectorized" code so that you have to skip here?

File: torch/_inductor/codegen/cpp.py (hunk 11 above)
Comment (jgong5) [reply torch/_inductor/codegen/cpp.py:1195]: Oh, for CPU, the `_load_mask` is initialized in the `ops.masked` call here: https://github.com/pytorch/pytorch/blob/6e3a7473cf45229a7a7b2e98611991f69d1c1437/torch/_inductor/codegen/cpp.py#L1204
It is not a no-op on CPU.

File: torch/_inductor/codecache.py [hunk 13]
```diff
@@ -893,6 +893,8 @@ def get_include_and_linking_paths(
             # and raise error together with instructions at compilation error later
         else:
             libs = ["omp"] if config.is_fbcode() else ["gomp"]
+            if config.assert_indirect_indexing:
+                libs += ["c10", "torch_cpu"]
```
Comment (lezcano) [inline torch/_inductor/codecache.py:897]: @jgong5 @peterbell10 @jansel The PR currently has some issues that we are not currently linking against `torchlib`, as the kernels cannot find `TORCH_CHECK`. Is it alright to unconditionally import these? This is in an `else:` block at the moment, but should probably be done for most configs if `assert_indirect_indexing`.

Comment (peterbell10) [reply torch/_inductor/codecache.py:897]: Not sure if we might run into ABI issues here, I'll just add that we only need `c10` for `TORCH_CHECK`.

Comment (lezcano) [reply torch/_inductor/codecache.py:897]: note that we already add `libs += ["c10_cuda", "cuda", "torch_cuda"]` for CUDA.

Comment (jgong5) [reply torch/_inductor/codecache.py:897]: > Is it alright to unconditionally import these?

Yes, sounds good to me.

File: torch/_inductor/codegen/cpp.py (hunk 12 above)
Comment (guilhermeleobas) [reply torch/_inductor/codegen/cpp.py:1417]: Peter, do you happen to know a test case or an example that requires a vectorized version of `TORCH_CHECK`?

Comment (lezcano) [reply torch/_inductor/codegen/cpp.py:1417]: See https://github.com/pytorch/pytorch/pull/108690#discussion_r1332273331

File: test/inductor/test_torchinductor.py (hunk 10 above)
Comment (guilhermeleobas) [reply test/inductor/test_torchinductor.py:811]: Negative indexing generates the following cpp kernel:


```cpp
/*
import torch

@torch.compile
def indexit(A, b):
    return A[-b]

A = torch.arange(20)
b = torch.zeros(4, dtype=torch.int)
print(indexit(A, b))
*/

extern "C" void kernel(const int* in_ptr0,
                       const long* in_ptr1,
                       long* out_ptr0)
{
    {
        for(long i0=static_cast<long>(0L); i0<static_cast<long>(4L); i0+=static_cast<long>(1L))
        {
            auto tmp0 = in_ptr0[static_cast<long>(i0)];
            auto tmp1 = decltype(tmp0)(-tmp0);
            auto tmp2 = tmp1 + 20;
            auto tmp3 = tmp1 < 0;
            auto tmp4 = tmp3 ? tmp2 : tmp1;
            TORCH_CHECK((0 <= tmp4) & (tmp4 < 20L), "index out of bounds: 0 <= tmp4 < 20L")
            auto tmp5 = in_ptr1[static_cast<long>(tmp4)];
            out_ptr0[static_cast<long>(i0)] = tmp5;
        }
    }
}
```

> @guilhermeleobas is your test generating "Vectorized" code so that you have to skip here?

Yes! Test `pos_and_neg` inside `test/inductor/test_torchinductor.py::test_neg_index` generates the following cpp code, to which contains `at::vec::Vectorize` instructions:

```cpp
extern "C" void kernel(const float* in_ptr0,
                       float* out_ptr0,
                       const long ks0)
{
    #pragma omp parallel num_threads(12)
    {
        {
            #pragma omp for
            for(long i0=static_cast<long>(0L); i0<static_cast<long>(8L*(at::native::div_floor_integer(((-2L) + ((-1L)*ks0)), (-8L)))); i0+=static_cast<long>(8L))
            {
                auto tmp0 = ([&]() { __at_align__ float tmpbuf[8]; for (long i0_inner = 0; i0_inner < 8; i0_inner++) tmpbuf[i0_inner] = in_ptr0[static_cast<long>(1L + ((-1L)*i0) + ((-1L)*i0_inner) + (((1L + ((-1L)*i0) + ((-1L)*i0_inner)) >= 0L) ? 0L : ks0))]; return at::vec::Vectorized<float>::loadu(tmpbuf); })();
                tmp0.store(out_ptr0 + static_cast<long>(i0));
            }
            #pragma omp for simd simdlen(4)
            for(long i0=static_cast<long>(8L*(at::native::div_floor_integer(((-2L) + ((-1L)*ks0)), (-8L)))); i0<static_cast<long>(at::native::div_floor_integer(((-2L) + ((-1L)*ks0)), (-1L))); i0+=static_cast<long>(1L))
            {
                auto tmp0 = in_ptr0[static_cast<long>(1L + ((-1L)*i0) + (((1L + ((-1L)*i0)) >= 0L) ? 0L : ks0))];
                out_ptr0[static_cast<long>(i0)] = tmp0;
            }
        }
    }
}
```

This happens because in this test, PyTorch doesn't call the `indirect_index` fallback here, and thus, no assertion is generated.
https://github.com/pytorch/pytorch/blob/c90f8c883d225ad9f25936f12b1cfcff23513cc1/torch/_inductor/index_propagation.py#L258-L262

File: test/inductor/test_torchinductor.py [hunk 14]
```diff
@@ -800,6 +800,74 @@ def repeat(x, n):
...
+        # Wrapping is constant-folded
+        test(flip_with_index_constant, (a,), has_assert=False, has_wrapping=False)
+
+        # Operation where we can't prove that the index is always positive or negative
+        def pos_and_neg(a):
+            b = torch.arange(start=1, end=-a.numel() - 1, step=-1, device=self.device)
+            return a[b]
+
+        # It has wrapping but no assert
+        test(pos_and_neg, (a,), has_assert=False, has_wrapping=(self.device == "cuda"))
```
Comment (lezcano) [inline test/inductor/test_torchinductor.py:858]: None of these tests should be data dependent. This piece of code should be just code movement

File: test/inductor/test_torchinductor.py (hunk 10 above)
Comment (lezcano) [reply test/inductor/test_torchinductor.py:811]: That path has wrapping but no assert, so that is fine, isn't it?

Now, we would like to have test cases where on CPU we test all the things we are testing in that test but also with vectorised loads/stores. In particular, we need to be able to have vectorised checks for it, and add the relevant code so that the whole thing can be vectorised. Without that, it could be the case that this PR is actually regressing and it's making some code that was previously vectorised, non-vectorised.

**Comment by lezcano** [comment]
The vectorised path is not being exercised. I get the feeling that this PR is disabling the vectorisation of some of these constructions, which is not good. Can you check whether this is the case, and if so fix it (we'll need to implement a TORCH_CHECK for the Vectorized class)

File: test/inductor/test_torchinductor.py (hunk 14 above)
Comment (guilhermeleobas) [reply test/inductor/test_torchinductor.py]: Fixed

**Comment by guilhermeleobas** [comment]
@lezcano, I just checked, and it doesn't seem the changes introduced here prevent vectorization.

**Review by lezcano — APPROVED** [review]
Thank you for the analysis!

Could you open issues for the following items:
- The bug you found
- The fact that we don't have a vectorised TORCH_CHECK
- The fact that we are not vectorising advanced indexing (depends on the previous point)

**Comment by lezcano** [comment]
you'll need to rebase

**Comment by izaitsevfb** [comment]
> @izaitsevfb who is going to land this one internally?

Please don't reopen yet, as the PR hasn't been reverted on GH. I'm trying to verify whether it could be re-landed internally, as the fix seems to be available.
